sdk/storage/inc/blobs/unused/blob_client.h

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+namespace Azure { namespace Storage {
+
+class BlobClient
+{
+public:
+    // connection string
+    BlobClient(const std::string& connectionString, std::string containerName, std::string blobName, BlobClientOptions options = BlobClientOptions());
+
+    // shared key auth
+    BlobClient(Azure::Core::Http::Uri uri, Azure::Storage::Common::StorageSharedKeyCredential credential, BlobClientOptions options = BlobClientOptions());
+
+    // token auth
+    BlobClient(Azure::Core::Http::Uri uri, Azure::Core::TokenCredential credential, BlobClientOptions options = BlobClientOptions());
+
+    // anonymous/SAS/customized pipeline auth
+    BlobClient(Azure::Core::Http::Uri uri, BlobClientOptions options = BlobClientOptions());
+
+    void SetSnapshotTime(const std::string& snapshot);
```

Comment:
> This makes the client no thread-safe. See the Go pattern way for doing this. Changing a snapshot (or version) returns a NEW client with the new URL. The only client remains so other threads using it are not affected.

In that case, we have to move these snapshot function and versionId function to derived class like BlockBlobClient, AppendBlobClient. Because in C++, we cannot override function only by return type.

Reply:
> You want the methods the base class in addition to the derived classes. The base class allows the customer to create a client without knowing what kind of blob it is. But this still should support snapshots and version.

> In c#, the base method can return the base type and the derived typed can return their own types. I'm not sure if c++can do it; we'd have to experiment and see what's possible.

Of course we want to do that, but that cannot be done in C++.

Reply:
Virtual methods would be nice to enable mocking - that's how we enable mocking in C#, for example. @BillyONeal : Maybe you have some ideas to help us here?

Reply:
Virtuals would force instances of this thing to be passed around by pointer and give up natural value semantics. I'm not saying that you can't do it but I don't recommend contorting the interface for that specific reason unless you expect to support your *customers* somehow using that as a mocking interface. (And then we need to define what the expected lifetime semantics will be)